source/TestAdapter/Executor.cs

```diff
@@ -586,143 +603,98 @@ private List<TestResult> RunTestOnEmulator(List<TestCase> tests)
-                // 2. mscorlib
-                // 3. test framework
-                // 4. test application
-                StringBuilder str = new StringBuilder();
-                foreach (var pe in allPeFiles)
-                {
-                    str.Append($" -load \"{Path.Combine(workingDirectory, pe)}\"");
-                }
-
-                string parameter = str.ToString();
+                arguments.Append($" \"{Path.Combine(workingDirectory, pe)}\"");
+            }
 
-                _logger.LogMessage(
-                    $"Parameters to pass to nanoCLR: <{parameter}>",
-                    Settings.LoggingLevel.Verbose);
+            // should we use a local nanoCLR instance?
+            if (!string.IsNullOrEmpty(_settings.PathToLocalCLRInstance))
+            {
+                arguments.Append($"  --localinstance \"{_settings.PathToLocalCLRInstance}\"");
+            }
 
-                var nanoClrLocation = TestObjectHelper.GetNanoClrLocation();
-                if (string.IsNullOrEmpty(nanoClrLocation))
-                {
-                    _logger.LogPanicMessage("Can't find nanoCLR Win32 in any of the directories!");
-                    results.First().Outcome = TestOutcome.Failed;
-                    results.First().ErrorMessage = "Can't find nanoCLR Win32 in any of the directories!";
-                    return results;
-                }
+            // if requested, set diagnostic output
+            if(_settings.Logging > Settings.LoggingLevel.None)
+            {
+                arguments.Append(" -v diag");
+            }
 
-                _logger.LogMessage($"Found nanoCLR Win32: {nanoClrLocation}", Settings.LoggingLevel.Verbose);
-                _nanoClr.StartInfo = new ProcessStartInfo(nanoClrLocation, parameter)
-                {
-                    WorkingDirectory = workingDirectory,
-                    UseShellExecute = false,
-                    RedirectStandardError = true,
-                    RedirectStandardOutput = true
-                };
+            _logger.LogMessage(
+                $"Launching nanoCLR with these arguments: '{arguments}'",
+                Settings.LoggingLevel.Verbose);
 
-                _logger.LogMessage(
-                    $"Launching process with nanoCLR (from {Path.GetFullPath(TestObjectHelper.GetNanoClrLocation())})",
-                    Settings.LoggingLevel.Verbose);
+            // launch nanoCLR
+            var cmd = Cli.Wrap("nanoclr")
+                 .WithArguments(arguments.ToString())
+                 .WithValidation(CommandResultValidation.None);
 
-                // launch nanoCLR
-                if (!_nanoClr.Start())
-                {
-                    results.First().Outcome = TestOutcome.Failed;
-                    results.First().ErrorMessage = "Failed to start nanoCLR";
+            // setup cancellation token with a timeout of 5 seconds
+            using (var cts = new CancellationTokenSource())
```

Comment:
minor, you can simplify by directly setting the time to cancel in the constructor